stdlib/public/Concurrency/Actor.cpp

```diff
@@ -988,6 +948,53 @@ static_assert(sizeof(ActiveActorStatus) == ACTIVE_ACTOR_STATUS_SIZE,
   "ActiveActorStatus is of incorrect size");
 #endif /* !SWIFT_CONCURRENCY_ACTORS_AS_LOCKS */
 
+class DefaultActorImplHeader : public HeapObject {
+protected:
+#if SWIFT_CONCURRENCY_ACTORS_AS_LOCKS
+  // If actors are locks, we don't need to maintain any extra bookkeeping in the
+  // ActiveActorStatus since all threads which are contending will block
+  // synchronously, no job queue is needed and the lock will handle all priority
+  // escalation logic
+  Mutex drainLock;
+#else
+  // Note: There is some padding that is added here by the compiler in order to
+  // enforce alignment. This is space that is available for us to use in
+  // the future
+  alignas(sizeof(ActiveActorStatus)) char StatusStorage[sizeof(ActiveActorStatus)];
+#endif
+  // TODO (rokhinip): Make this a flagset
+  bool isDistributedRemoteActor;
+};
+
+// All the fields accessed under the actor's lock should be moved
+// to the end of the default-actor reservation to minimize false sharing.
+// The memory following the DefaultActorImpl object are the stored properties of
+// the actor, which are likely all isolated and therefore used only by the
+// current processing thread.
+class DefaultActorImplFooter {
+protected:
+#if !SWIFT_CONCURRENCY_ACTORS_AS_LOCKS
+  using SimpleQueue = swift::SimpleQueue<Job *, JobQueueTraits>;
+  using PriorityQueue = swift::PriorityQueue<Job *, JobQueueTraits>;
+
+  // When enqueued, jobs are atomically added to a linked list with the head
+  // stored inside ActiveActorStatus. This list contains jobs in the LIFO order
+  // regardless of their priorities.
+  //
+  // After that processing thread collects them and adds to the
+  // `prioritisedJobs` while holding the actor lock.
+  //
+  PriorityQueue prioritisedJobs;
+
+  // As an optimisation, we handle unprioritized jobs when obtaining actor lock.
+  // This flag is used to skip `processIncomingQueue()` on the iteration of the
+  // drainer loop right after taking the lock.
+  // This applies both to obtaining lock when entering the drainer loop, and
+  // when assuming thread for switching to a new actor.
```

Comment:
Hmm.  Why does this have to be persistent state on the actor instead of something we just do locally in `defaultActorDrain` whenever it's not our first trip through the loop?

Reply:
This is needed to support `tryLock(false)`, which can also happen inside `runJobInEstablishedExecutorContext()` if job calls `swift_task_switch()` - so it is not just first trip through the loop, but rather any trip through the loop right after `tryLock()`. If we don't want to optimise atomic operations  for `tryLock(false)`, then this flag is not needed at all - see commit history.

Alternatively it could live in `ExecutorTrackingInfo`. Currently it is accessed only in methods of the `DefaultActorImpl`, so accessing it as a member variable is rather convenient. But if we want to be frugal about available memory in `DefaultActorImpl`, it can be moved to `ExecutorTrackingInfo`.

Reply:
I think you may be misunderstanding how actor switching works.  In the happy path, if we're currently on actor A and switching to actor B, we just lock B, unlock A, and immediately return to the task.  We don't go back to whatever outer processing loop we might be in until the current task suspends or completes.

Reply:
My understanding goes as following:
* `defaultActorDrain()` starts with actor `A` as `currentActor`.
* job isolated to actor `A`, calls `swift_task_switch()`
* assuming happy path is taken - actor `B` is locked, actor `A` is unlocked.
* when locking actor `B`, `handleUnprioritizedJobs()` is called
* if task suspends or completes without more actor switches, execution returns to `defaultActorDrain()`, but now with actor `B` being the `currentActor`.
* the same thread that was draining actor `A` is now draining actor `B`.
* since we've already called `handleUnprioritizedJobs()`, we don't want to call it again before dequeueing a job from `prioritizedJobs`.

Is anything incorrect here?

All of this applies only if we want to support `handleUnprioritizedJobs()` in `tryLock(false)`. I was curious to explore this option, but I don't insist on it. If we don't need it - I'll just revert two commits from the history.

Reply:
The only thing you're missing is that an arbitrary amount of time could elapse between locking actor `B` and the current task suspending — we've returned control to user code in the meantime.

Reply:
Ok, I’ll revert two commits adding support for `handleUnprioritizedJobs()` in `tryLock(false)` and add a comment explaining the reasoning behind it.